### ISSUE TITLE
persistent storage of sbc intermediate results, feature proposal

### ISSUE THREAD

**Issue by jpritikin** [stan-dev/rstan#641]
If R crashes for whatever reason during sbc then all work is lost.

Would you accept a PR to add a save_progress="/path/to/directory" argument that would save each model run as they finish and then restore them if available?

**Comment by jgabry** [comment]
Something like that seems like a good idea. @bgoodri what do you think?

**Comment by bgoodri** [comment]
I guess. It seems as if sbc() is crashing R, we have bigger problems.
Continuing an interrupted run is tricky with the PRNG seeds and such, but
it may be possible.

**Comment by jpritikin** [comment]
> It seems as if sbc() is crashing R, we have bigger problems.

It's not that. It's stuff like me installing a new version of rstan while I'm running sbc or needing to reboot my laptop for some other reason. I have a model that takes 4-5 hours to sample 100 models with sbc. That's just a long time, and I have another model that is going to take even longer.

> Continuing an interrupted run is tricky with the PRNG seeds

There is already code in sbc that takes care to set the seeds such that sbc runs are reproducible.

**Comment by jpritikin** [comment]
I got some RStudio crashes yesterday. That's not your fault, but interferes with SBC.

**Comment by jpritikin** [comment]
With 150 posteriors, I'm running out of memory on my laptop with 16G of RAM. It looks like we need to change how data are loaded from disk. Instead of loading everything into `post`, we need to load one just to find the structure and then load one at a time, throwing away the data that we don't use.

**Comment by jpritikin** [comment]
Hey, can we summarize the rank indicators using colSums as we read in each posterior? That could save a lot of RAM, but the thinning would have to happen at this point.

**Comment by jpritikin** [comment]
Ah, here's where the memory is going,

```
> object_size(fit)
93.1 MB
> object_size(fit@stanmodel)
91.1 MB
> fake <- list()
> class(fake) <- 'stanmodel'
> fit@stanmodel <- fake
> object_size(fit)
6.2 MB
```
Looks like this is the simplest solution.

**Comment by bob-carpenter** [comment]
Yes.  If we code up (theta < theta_sim) as an indicator in generated quantities, then the sum gives the rank and that's all you need to store for SBC.

**Comment by jpritikin** [comment]
> Yes. If we code up (theta < theta_sim) as an indicator in generated quantities, then the sum gives the rank and that's all you need to store for SBC.

Once you summarize as rank sums then you can't change the thinning.

Most of the memory was getting consumed by the stanmodel. Not saving the model to disk solved my problem. I don't think it's a big deal to store the unsummarized, per-draw rankings.

**Comment by bob-carpenter** [comment]
When I built this I used iterative deepening that kept doubling the iterations and doubling the thinning period until the effective sample size rate was over 90% of the total draws.  Another approach would be to save more than you think you need, measure ESS, and thin based on autocorrelation rate.  Either way, you only ever need to keep one chain's output in memory.

**Comment by jpritikin** [comment]
Sure, I'm strongly in favor of automating the choice of thinning. I'm just saying that we can't justify it by the savings in storage requirements. If we can improve the user experience, we should do that.

**Comment by jgabry** [comment]
I think we want to avoid the case where someone needs to rerun SBC because they only have the already thinned output but they want to choose a different thinning interval. It’s true that returning the indicators pre summation will result in much bigger objects but that seems preferable to possibly having to rerun everything. Or am I misunderstanding the issue?
